src/Microsoft.AspNet.Security.OAuth/Notifications/OAuthBearerAuthenticationNotifications.cs

```diff
@@ -3,64 +3,61 @@
-            OnValidateIdentity = context => Task.FromResult<object>(null);
-            OnApplyChallenge = context =>
-            {
-                context.HttpContext.Response.Headers.AppendValues("WWW-Authenticate", context.Challenge);
-                return Task.FromResult(0);
-            };
-        }
-
-        /// <summary>
-        /// Handles processing OAuth bearer token.
-        /// </summary>
-        public Func<OAuthRequestTokenContext, Task> OnRequestToken { get; set; }
+			AuthenticationFailed = notification => Task.FromResult(0);
+			MessageReceived = notification => Task.FromResult(0);
+			SecurityTokenReceived = notification => Task.FromResult(0);
+			SecurityTokenValidated = notification => Task.FromResult(0);
 
-        /// <summary>
-        /// Handles validating the identity produced from an OAuth bearer token.
-        /// </summary>
-        public Func<OAuthValidateIdentityContext, Task> OnValidateIdentity { get; set; }
+			OnApplyChallenge = context =>
+            {
+				context.HttpContext.Response.Headers.AppendValues("WWW-Authenticate", context.Challenge);
+				return Task.FromResult(0);
+			};
+		}
 
-        /// <summary>
+		/// <summary>
         /// Handles applying the authentication challenge to the response message.
         /// </summary>
         public Func<OAuthChallengeContext, Task> OnApplyChallenge { get; set; }
 
-        /// <summary>
-        /// Handles processing OAuth bearer token.
-        /// </summary>
-        /// <param name="context"></param>
-        /// <returns></returns>
-        public virtual Task RequestToken(OAuthRequestTokenContext context)
-        {
-            return OnRequestToken(context);
-        }
+		/// <summary>
+		/// Invoked if exceptions are thrown during request processing. The exceptions will be re-thrown after this event unless suppressed.
+		/// </summary>
+		public Func<AuthenticationFailedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>, Task> AuthenticationFailed { get; set; }
 
-        /// <summary>
-        /// Handles validating the identity produced from an OAuth bearer token.
-        /// </summary>
-        /// <param name="context"></param>
-        /// <returns></returns>
-        public virtual Task ValidateIdentity(OAuthValidateIdentityContext context)
-        {
-            return OnValidateIdentity.Invoke(context);
-        }
+		/// <summary>
+		/// Invoked when a protocol message is first received.
+		/// </summary>
+		public Func<MessageReceivedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>, Task> MessageReceived { get; set; }
 
-        /// <summary>
+		/// <summary>
+		/// Invoked with the security token that has been extracted from the protocol message.
+		/// </summary>
+		public Func<SecurityTokenReceivedNotification<OAuthRequestTokenContext, OAuthBearerAuthenticationOptions>, Task> SecurityTokenReceived { get; set; }
```

Comment:
Your reworked generic `SecurityTokenReceivedNotification` is used in a weird way: I'm not sure that `OAuthRequestTokenContext` can semantically be considered as a `ProtocolMessage`.